### ISSUE TITLE
Add associated types and effects to the Flix Standard Library

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#7433]
Post-merge of the new typer:

Ordered from simple to more complex:

- [ ] Add an associated type to `Add` to allow `Set#{1, 2, 3} + 1`.
- [ ] Add an associated type to `Sub` to allow `Set#{1, 2, 3} - 1`.
- [ ] Add an associated type to `Coerce`.
- [ ] Add associated types and effects to `ToJava` and `FromJava`.
- [ ] Add associated types and effects to `Iterable`. Implement new instances for mutable collections.
- [ ] Add an associated effect to `ToString`. Implement new instances for mutable collections.
- [ ] Add an associated effect to `Foldable`. Implement new instances for mutable collections.
- [ ] Add an associated effect to `Reducible`. Implement new instances for mutable collections.

To be discussed:

- Add associated effects to `Add`, `Sub`, ... `Div`...?
- Add associated effects to `Functor` and friends...?
- Add associated effects to `Filterable`, `Traversable`, and friends...?

**Comment by magnus-madsen** [comment]
@stephentetley Would you be interested in leading this? (There is no rush)

We could use someone with a steady hand to ensure that we keep the standard library consistent.

**Comment by stephentetley** [comment]
Yes I can do, I'm a bit busy for a week or so as I've some work related (not Flix) code that I want to move forward plus finishing off the `class` to `trait` and `toVector`. But after that I'll be on it.

**Comment by stephentetley** [comment]
`UnorderedFoldable` is also to do.
